File: src/graphs/algorithms.py
from __future__ import annotations

import heapq
from collections import deque

from src.graphs.graph import Graph


def bfs(graph: Graph, source: str) -> dict[str, dict]:
    result: dict[str, dict] = {source: {"level": 0, "parent": None}}
    queue: deque[str] = deque([source])

    while queue:
        u = queue.popleft()
        for v, _ in graph.neighbors(u):
            if v not in result:
                result[v] = {"level": result[u]["level"] + 1, "parent": u}
                queue.append(v)

    return result


def dfs(graph: Graph, source: str) -> dict[str, dict]:
    result: dict[str, dict] = {}
    edge_types: list[tuple[str, str, str]] = []
    time_counter = 0

    def _dfs_iterative(start: str) -> None:
        nonlocal time_counter
        time_counter += 1
        result[start] = {"discovery": time_counter, "finish": 0, "parent": None}
        stack: list[tuple[str, int]] = [(start, 0)]

        while stack:
            u, idx = stack[-1]
            neighbors = graph.neighbors(u)

            if idx < len(neighbors):
                stack[-1] = (u, idx + 1)
                v, _ = neighbors[idx]

                if v not in result:
                    edge_types.append((u, v, "tree"))
                    time_counter += 1
                    result[v] = {"discovery": time_counter, "finish": 0, "parent": u}
                    stack.append((v, 0))
                elif result[v]["finish"] == 0:
                    if v != result[u]["parent"] or graph.directed:
                        edge_types.append((u, v, "back"))
                elif result[v]["discovery"] > result[u]["discovery"]:
                    edge_types.append((u, v, "forward"))
                else:
                    edge_types.append((u, v, "cross"))
            else:
                time_counter += 1
                result[u]["finish"] = time_counter
                stack.pop()

    _dfs_iterative(source)

    for node in graph.nodes():
        if node not in result:
            _dfs_iterative(node)

    has_cycle = any(t == "back" for _, _, t in edge_types)
    return {"nodes": result, "edge_types": edge_types, "has_cycle": has_cycle}


def dijkstra(
    graph: Graph, source: str
) -> tuple[dict[str, float], dict[str, str | None]]:
    for u in graph.nodes():
        for _, w in graph.neighbors(u):
            if w < 0:
                raise ValueError(
                    f"Dijkstra nao suporta pesos negativos (encontrado peso {w})"
                )

    dist: dict[str, float] = {source: 0.0}
    prev: dict[str, str | None] = {source: None}
    heap: list[tuple[float, str]] = [(0.0, source)]

    while heap:
        d, u = heapq.heappop(heap)
        if d > dist.get(u, float("inf")):
            continue
        for v, w in graph.neighbors(u):
            nd = d + w
            if nd < dist.get(v, float("inf")):
                dist[v] = nd
                prev[v] = u
                heapq.heappush(heap, (nd, v))

    return dist, prev


def bellman_ford(
    graph: Graph, source: str
) -> tuple[dict[str, float], dict[str, str | None], bool]:
    nodes = graph.nodes()
    dist: dict[str, float] = {n: float("inf") for n in nodes}
    prev: dict[str, str | None] = {n: None for n in nodes}
    dist[source] = 0.0

    all_edges: list[tuple[str, str, float]] = []
    for u in nodes:
        for v, w in graph.neighbors(u):
            all_edges.append((u, v, w))

    for _ in range(len(nodes) - 1):
        updated = False
        for u, v, w in all_edges:
            if dist[u] + w < dist[v]:
                dist[v] = dist[u] + w
                prev[v] = u
                updated = True
        if not updated:
            break

    has_negative_cycle = False
    for u, v, w in all_edges:
        if dist[u] + w < dist[v]:
            has_negative_cycle = True
            break

    return dist, prev, has_negative_cycle


def reconstruct_path(
    prev: dict[str, str | None], source: str, target: str
) -> list[str] | None:
    if target not in prev or (prev[target] is None and target != source):
        return None
    path: list[str] = []
    current: str | None = target
    while current is not None:
        path.append(current)
        current = prev[current]
    path.reverse()
    if path[0] != source:
        return None
    return path

File: web/app/lib/algorithms.ts
import { Graph } from "./graph";
import { BFSResult, DFSResult } from "./types";

export function bfs(graph: Graph, source: string): BFSResult {
  const result: BFSResult = { [source]: { level: 0, parent: null } };
  const queue: string[] = [source];
  while (queue.length > 0) {
    const u = queue.shift()!;
    for (const [v] of graph.neighbors(u)) {
      if (!(v in result)) {
        result[v] = { level: result[u].level + 1, parent: u };
        queue.push(v);
      }
    }
  }
  return result;
}

export function dfs(graph: Graph, source: string): DFSResult {
  const nodes: DFSResult["nodes"] = {};
  const edgeTypes: [string, string, string][] = [];
  let time = 0;

  function visit(start: string) {
    time++;
    nodes[start] = { discovery: time, finish: 0, parent: null };
    const stack: [string, number][] = [[start, 0]];

    while (stack.length > 0) {
      const [u, idx] = stack[stack.length - 1];
      const nbrs = graph.neighbors(u);
      if (idx < nbrs.length) {
        stack[stack.length - 1][1]++;
        const [v] = nbrs[idx];
        if (!(v in nodes)) {
          edgeTypes.push([u, v, "tree"]);
          time++;
          nodes[v] = { discovery: time, finish: 0, parent: u };
          stack.push([v, 0]);
        } else if (nodes[v].finish === 0) {
          if (v !== nodes[u].parent) edgeTypes.push([u, v, "back"]);
        } else if (nodes[v].discovery > nodes[u].discovery) {
          edgeTypes.push([u, v, "forward"]);
        } else {
          edgeTypes.push([u, v, "cross"]);
        }
      } else {
        time++;
        nodes[u].finish = time;
        stack.pop();
      }
    }
  }

  visit(source);
  for (const n of graph.nodes()) {
    if (!(n in nodes)) visit(n);
  }

  return {
    nodes,
    edgeTypes,
    hasCycle: edgeTypes.some(([, , t]) => t === "back"),
  };
}

export function dijkstra(
  graph: Graph,
  source: string
): { dist: Record<string, number>; prev: Record<string, string | null> } {
  const dist: Record<string, number> = {};
  const prev: Record<string, string | null> = {};
  const visited = new Set<string>();

  for (const n of graph.nodes()) {
    dist[n] = Infinity;
    prev[n] = null;
  }
  dist[source] = 0;

  // Simple priority queue via sorted array (graph is small)
  const heap: [number, string][] = [[0, source]];

  while (heap.length > 0) {
    heap.sort(([a], [b]) => a - b);
    const [d, u] = heap.shift()!;
    if (visited.has(u)) continue;
    visited.add(u);
    if (d > dist[u]) continue;

    for (const [v, w] of graph.neighbors(u)) {
      const nd = d + w;
      if (nd < dist[v]) {
        dist[v] = nd;
        prev[v] = u;
        heap.push([nd, v]);
      }
    }
  }

  return { dist, prev };
}

export function bellmanFord(
  graph: Graph,
  source: string
): { dist: Record<string, number>; prev: Record<string, string | null>; hasNegativeCycle: boolean } {
  const nodes = graph.nodes();
  const dist: Record<string, number> = Object.fromEntries(nodes.map((n) => [n, Infinity]));
  const prev: Record<string, string | null> = Object.fromEntries(nodes.map((n) => [n, null]));
  dist[source] = 0;

  const allEdges = graph.edges();

  for (let i = 0; i < nodes.length - 1; i++) {
    let updated = false;
    for (const [u, v, w] of allEdges) {
      if (dist[u] + w < dist[v]) {
        dist[v] = dist[u] + w;
        prev[v] = u;
        updated = true;
      }
      if (dist[v] + w < dist[u]) {
        dist[u] = dist[v] + w;
        prev[u] = v;
        updated = true;
      }
    }
    if (!updated) break;
  }

  let hasNegativeCycle = false;
  for (const [u, v, w] of allEdges) {
    if (dist[u] + w < dist[v]) { hasNegativeCycle = true; break; }
  }

  return { dist, prev, hasNegativeCycle };
}

export function reconstructPath(
  prev: Record<string, string | null>,
  source: string,
  target: string
): string[] | null {
  if (!(target in prev) || (prev[target] === null && target !== source)) return null;
  const path: string[] = [];
  let current: string | null = target;
  while (current !== null) {
    path.push(current);
    current = prev[current];
  }
  path.reverse();
  if (path[0] !== source) return null;
  return path;
}
